code/lib/codemod/src/transforms/__tests__/mdx-to-csf.test.ts

```diff
@@ -0,0 +1,412 @@
+
+  expect(mdx).toMatchInlineSnapshot(`
+    import { Meta, Story } from '@storybook/blocks';
+    import * as FoobarStories from './Foobar.stories';
+
+    <Meta of={FoobarStories} />
+
+    <Story id="button--primary" />
+
+  `);
+});
+
+test('convert correct story nodes', () => {
+  const input = dedent`
+      import { Meta, Story } from '@storybook/addon-docs';
+
+      <Meta title="Foobar" />
+      
+      <Story name="Primary">Story</Story>
+    `;
+
+  const mdx = jscodeshift({ source: input, path: 'Foobar.stories.mdx' });
+
+  expect(mdx).toMatchInlineSnapshot(`
+    import { Meta, Story } from '@storybook/blocks';
+    import * as FoobarStories from './Foobar.stories';
+
+    <Meta of={FoobarStories} />
+
+    <Story of={FoobarStories.Primary} />
+
+  `);
+});
+
+test('convert story nodes with spaces', () => {
+  const input = dedent`
+      import { Meta, Story } from '@storybook/addon-docs';
+
+      <Meta title="Foobar" />
+      
+      <Story name="Primary Space">Story</Story>
+    `;
+
+  const mdx = jscodeshift({ source: input, path: 'Foobar.stories.mdx' });
+
+  expect(mdx).toMatchInlineSnapshot(`
+    import { Meta, Story } from '@storybook/blocks';
+    import * as FoobarStories from './Foobar.stories';
+
+    <Meta of={FoobarStories} />
+
+    <Story of={FoobarStories.PrimarySpace} />
+
+  `);
+
+  const [, csf] = fs.writeFileSync.mock.calls[0];
+  expect(csf).toMatchInlineSnapshot(`
+    export default {
+      title: 'Foobar',
+    };
+
+    export const PrimarySpace = {
+      render: () => 'Story',
+      name: 'Primary Space',
+    };
+
+  `);
+});
+
+test('extract esm into csf head code', () => {
+  const input = dedent`
+      import { Meta, Story } from '@storybook/addon-docs';
+
+      # hello
+
+      export const args = { bla: 1 };
+      
+      <Meta title="foobar" />
+
+      world {2 + 1}
+
+      <Story name="foo">bar</Story>
+      
+      <Story 
+        name="Unchecked"
+        args={{
+          ...args,
+          label: 'Unchecked',
+        }}>
+        {Template.bind({})}
+      </Story>
+    `;
+
+  const mdx = jscodeshift({ source: input, path: 'Foobar.stories.mdx' });
+
+  expect(mdx).toMatchInlineSnapshot(`
+    import { Meta, Story } from '@storybook/blocks';
+    import * as FoobarStories from './Foobar.stories';
+
+    # hello
+
+    export const args = { bla: 1 };
+
+    <Meta of={FoobarStories} />
+
+    world {2 + 1}
+
+    <Story of={FoobarStories.Foo} />
+
+    <Story of={FoobarStories.Unchecked} />
+
+  `);
+
+  const [csfFileName, csf] = fs.writeFileSync.mock.calls[0];
+  expect(csfFileName).toMatchInlineSnapshot(`Foobar.stories.js`);
+  expect(csf).toMatchInlineSnapshot(`
+      const args = { bla: 1 };
+
+      export default {
+        title: 'foobar',
+      };
+
+      export const Foo = {
+        render: () => 'bar',
+        name: 'foo',
+      };
+
+      export const Unchecked = {
+        render: Template.bind({}),
```

Comment:
Template is undefined here (but was undefined in the MDX file too). Are we only bringing variable references across if they actually exist?

Reply:
Good point, I made the test correct. We don't check if references actually exist, but we do make sure to copy all ESM code from the mdx file to the CSF file, so all (and maybe more) references should exist.